### PR TITLE
WIP feat(routes): add support for route middleware

### PR DESCRIPTION
**Pull request by hypeJunction**
Plugins can now define middleware to be executed before route is rendered.
Makes it simpler to allow plugins to define public routes available
when walled garden is enabled.

- [ ] Needs #11687

### CODE REVIEW

File: engine/classes/Elgg/WalledGatdenException.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Elgg;
+
+/**
+ * Thrown when walled garden gatekeeper prevents access
+ */
+class WalledGatdenException extends GatekeeperException {
```
Comment (hypeJunction) [inline engine/classes/Elgg/WalledGatdenException.php]: Garden that is

File: docs/guides/routing.rst [hunk 2]
```diff
@@ -162,6 +162,59 @@ By default, Elgg will set the following requirements for named URL segments:
...
+			],
+		]
+	];
+
+You can additionally specify middleware, which will be executed before the resource is rendered. Middleware can be used
+to prevent access to a route by throwing an ``HttpException``, as well as to perform other actions (e.g. logging analytics).
+Core provides 3 middleware handlers: ``Gatekeeper``, ``AdminGatekeeper``, ``WalledGarden``. Plugins can implement their
+own middleware by extending ``\Elgg\Router\Middleware`` class.
+
+.. php::code
```
Comment (jeabakker) [inline docs/guides/routing.rst]: almost should be `.. code-block:: php`

File: engine/classes/Elgg/Router.php [hunk 3]
```diff
@@ -260,9 +289,13 @@ public function registerPageHandler($identifier, $function) {
 	 */
 	public function registerRoute($name, array $params = []) {
 
+		$params = $this->hooks->trigger('route:config', $name, $params, $params);
+
 		$path = elgg_extract('path', $params);
 		$resource = elgg_extract('resource', $params);
 		$handler = elgg_extract('handler', $params);
+		$middleware = elgg_extract('middleware', $params, []);
+		$public = elgg_extract('public', $params, false);
```
Comment (jeabakker) [inline engine/classes/Elgg/Router.php]: why default to non-public routes, doesn't that cause lots of problems

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Router.php (hunk 3 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Router.php]: That's the current approach. When you enable WG everything is non-public, unless allowed by hook

Comment (jeabakker) [reply engine/classes/Elgg/Router.php]: ah so this only gets enforced by WG (and added middleware)?

Comment (hypeJunction) [reply engine/classes/Elgg/Router.php]: I suppose "public" is ambiguous. Should rename into something else. Ideas?

**Comment by jeabakker** [comment]
Maybe "guarded" or "gated" I don't have anything better

Op 31 jan. 2018 16:36 schreef "Ismayil Khayredinov" <
notifications@github.com>:

**Comment by hypeJunction** [comment]
Fixes #11177

**Comment by hypeJunction** [comment]
We will need to eventually kill context and input altogether. Both are stupid and volatile, but this will do for now. Request will need to be thrown out of services as it is not a service, it needs to be property of the router after being routed. Route registration needs to move to it's own service.

**Comment by hypeJunction** [comment]
Once this is in, there is another PR coming with rewiring actions system to use routes and middleware. Almost done there, just need to update the tests. And we will finally have a normal uniform OOP routing system

**Comment by hypeJunction** [comment]
Replaced by #11690

**Pull request closed without merging**